### PR TITLE
Ittest improve

### PR DESCRIPTION
**Pull request by stonedu1011**
## Description

`ittest` has being a very useful tool to test components that communicate with external servers via HTTP.  However, `ittest` was designed to use with `bootstrap` and `apptest` and cannot be used as a standalone test utility. 

The purpose of this PR is to improve `ittest` for standalone usage. Projects that doesn't use `bootstrap`, `apptest` or `uber/fx` can still leverage utilities provided by `ittest` package.

## Type of Change

- [ ] Bug Fix
- [X] New Feature
- [ ] Breaking Change
- [X] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [X] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [X] Existing issues have been referenced (where applicable)
- [X] I have verified this change is not present in other open pull requests
- [X] Functionality is documented
- [X] All code style checks pass
- [X] New code contribution is covered by automated tests
- [X] All new and existing tests pass

### CODE REVIEW

File: test/ittest/httpvcr_test.go [hunk 1]
```diff
@@ -107,104 +107,144 @@ func (c TestController) Post(_ context.Context, r *TestRequest) (*TestResponse,
...
+		webtest.WithRealServer(),
+		test.Setup(SetupVCR(
+			HttpRecordingMode(),
+			HttpRecordName(RecordAltName),
+			HttpRecorderHooks(LocalhostRewriteHook()),
+			SanitizeHttpRecord(),
+			FixedHttpRecordDuration(DefaultHTTPDuration),
+			FixedHttpRecordDuration(0),
+			FixedHttpRecordDuration(DefaultHTTPDuration),
+			HttpTransport(http.DefaultTransport),
```
Comment (TimShi) [inline test/ittest/httpvcr_test.go:221]: developer doesn't need to comment out this line when going from recording mode to playback mode, right? 

also, I think the `recorder` package already sets it to this value, so this line is optional, so this line is optional?

File: test/ittest/httpvcr.go [hunk 2]
```diff
@@ -87,41 +87,66 @@ func WithHttpPlayback(t *testing.T, opts ...HTTPVCROptions) test.Options {
...
+func ResetRecorder(ctx context.Context) {
+	rec, ok := ctx.Value(ckRecorder).(*HttpRecorder)
+	if !ok || rec.Recorder == nil {
+		return
+	}
+	rec.SetMatcher(rec.Matcher)
+}
+
+// StopRecorder stops the recorder extracted from the given context.
+func StopRecorder(ctx context.Context) error {
```
Comment (TimShi) [inline test/ittest/httpvcr.go:142]: Probably should add a comment on these methods that they are only required to be used in tests without using the test bootstrap.

**Comment by TimShi** [comment]
We should probably add a readme on what features we added on top of recorder and casset, and how it can be used independent of the test bootstrap.

**Review by TimShi — APPROVED** [review]

File: test/ittest/httpvcr.go [hunk 3]
```diff
@@ -182,7 +207,32 @@ func HttpRecordMatching(opts ...RecordMatcherOptions) HTTPVCROptions {
...
+}
+
+// DisableHttpRecorderHooks returns a HTTPVCROptions that removes installed hooks by name
+func DisableHttpRecorderHooks(names ...string) HTTPVCROptions {
+	return func(opt *HTTPVCROption) {
+	LOOP:
+		for i := range names {
+			for j := range opt.Hooks {
+				if names[i] == opt.Hooks[j].Name() {
+					opt.Hooks[j] = opt.Hooks[len(opt.Hooks)-1]
```
[the remark this answers is not in the document]
Comment (stonedu1011) [reply test/ittest/httpvcr.go:230]: hooks are sorted when creating the recorder based on `OrderedFirstCompare` rule, see `NewHttpRecorder()`

File: test/ittest/httpvcr.go [hunk 4]
```diff
@@ -192,17 +242,29 @@ func HttpRecordIgnoreHost() HTTPVCROptions {
 	return HttpRecordMatching(IgnoreHost())
 }
 
+// HttpRecordOrdering toggles HTTP interactions order matching.
+// When enforced, HTTP interactions have to happen in the recorded order.
+// Otherwise, HTTP interactions can happen in any order, but each matched record can only replay once
+// By default, record ordering is enabled
+func HttpRecordOrdering(enforced bool) HTTPVCROptions {
```
[the remark this answers is not in the document]
Comment (stonedu1011) [reply test/ittest/httpvcr.go:249]: This is a new option. We only had `DisableHttpRecordOrdering` before. The reason `opensearchtest` need to update is due to the default behaviour change of `NewHttpRecorder()`

File: test/ittest/httpvcr.go [hunk 5]
```diff
@@ -215,76 +277,105 @@ func ApplyHttpLatency() HTTPVCROptions {
...
  *************************/
 
-// HttpRecorder wrapper of recorder.Recorder, used to hold some value that normally inaccessible via wrapped recorder.Recorder.
-// Note: Internal Use Only! this type is for other test utilities to re-configure recorder.Recorder
+// HttpRecorder wrapper of recorder.RawRecorder, used to hold some value that normally inaccessible via wrapped recorder.RawRecorder.
+// Note: This type is for other test utilities to re-configure recorder.RawRecorder
 type HttpRecorder struct {
 	*recorder.Recorder
 	RawOptions *recorder.Options
 	Matcher    cassette.MatcherFunc
```
[the remark this answers is not in the document]
Comment (stonedu1011) [reply test/ittest/httpvcr.go:352]: changed it to `InitMatcher`

File: test/ittest/httpvcr_test.go (hunk 1 above)
Comment (stonedu1011) [reply test/ittest/httpvcr_test.go:221]: developers do need to comment out `HttpRecordingMode()` to switch from recording mode to playback mode. The default is playback. 

`FixedHttpRecordDuration()` is a "recording" specific option. It takes no effect in playback mode. The playback counterpart options is `ApplyHttpLatency()`

I added an comment on the option method to clarify that.

**Pull request merged**